doc/texttopng.1
.\"
.\" Copyright (C) 2011 Richard Kettlewell
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.\"
.TH texttopng 1
.SH NAME
texttopng \- convert text to PNG
.SH SYNOPSIS
\fBtexttopng\fR [\fIOPTIONS\fR] [\fB--\fR] [\fIINPUT\fR ...]
.SH DESCRIPTION
\fBtexttopng\fR reads one or more text files and converts them to PNG images.
.PP
If there are no \fIINPUT\fR arguments then input is read from standard
input.
Otherwise all the named input files are concatenated.
.PP
PNGs are written to \fBoutput0.png\fR, \fBoutput1.png\fR, etc.
The base name can be changed with the \fB-o\fR option.
.PP
The use of overprinting to represent bold and underline, for instance
as traditionally found in \fBman\fR(1) output, is recognized.
.SH OPTIONS
.SS "Text Options"
.TP
.B -t\fR, \fB--tab \fICOLUMNS
Set the size of a tab.
The default is 8 columns.
.TP
.B -f\fR, \fB--font-size \fISIZE
Set the font size.
The default is 10.
.TP
.B -F\fR, \fB--font \fIFONT
Set the font.
The default is \fBCourier New\fR.
.TP
.B -i\fR, \fB--italic
Render underlined text in italics.
.SS "Layout Options"
.TP
.B -w\fR, \fB--width \fIPIXELS
Set the width of the output image.
The default is 80 columns.
.TP
.B -H\fR, \fB--height \fIPIXELS
Set the height of the output image.
The default is 66 rows.
.TP
.B -b\fR, \fB--border \fIPIXELS
Set the border size.
The default is no border.
.TP
.B -T\fR, \fB--title \fITITLE
Add a title to each page.
.TP
.B -p\fR, \fB--page-numbering
Enable page numbering.
The border must be large enough for the page number.
.SS "Other Options"
.TP
.B -l\fR, \fB--list-fonts
List the known fonts.
.TP
.B -o\fR, \fB--output \fIBASE
Set the base name for the output files.
Output will be written to \fIBASE\fB0.png\fR, \fIBASE\fB1.png\fR, etc.
.TP
.B -h\fR, \fB--help
Display a usage message.
.TP
.B -V\fR, \fB--version
Display the version string.


src/texttopng.cc
// 
// Copyright (C) 2011 Richard Kettlewell
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
// 
#include <config.h>
#include "CairoOutput.hh"
#include "Utils.hh"
#include <cerrno>
#include <cstring>
#include <cstdlib>
#include <getopt.h>
#include <pangomm/init.h>

static const struct option options[] = {
  { "tab", required_argument, NULL, 't' },
  { "width", required_argument, NULL, 'w' },
  { "height", required_argument, NULL, 'H' },
  { "font-size", required_argument, NULL, 'f' },
  { "font", required_argument, NULL, 'F' },
  { "output", required_argument, NULL, 'o' },
  { "list-fonts", no_argument, NULL, 'l' },
  { "border", required_argument, NULL, 'b' },
  { "page-numbering", no_argument, NULL, 'p' },
  { "title", required_argument, NULL, 'T' },
  { "italic", no_argument, NULL, 'i' },
  { "help", no_argument, NULL, 'h' },
  { "version", no_argument, NULL, 'V' },
  { NULL, 0, NULL, 0 },
};

static void help() {
  printf("Usage:\n"
         "  texttopng [OPTIONS] [--] [INPUT...]\n"
         "  texttopng --list-fonts\n"
         "\n"
         "Text options:\n"
         "  -t, --tab COLUMNS           Set tab size (default=8)\n"
         "  -f, --font-size SIZE        Set font size\n"
         "  -F, --font FONT             Set font (default=Courier New)\n"
         "  -i, --italic                Render underlined text in italic\n"
         "Layout options:\n"
         "  -w, --width PIXELS          Set output file width\n"
         "  -H, --height PIXELS         Set output file height\n"
         "  -b, --border PIXELS         Set border\n"
         "  -T, --title TITLE           Top-of-page title string\n"
         "  -p, --page-numbering        Enable page numbering\n"
         "Other options:\n"
         "  -l, --list-fonts            List fonts\n"
         "  -o, --output BASE           Set output base (default=output)\n"
         "  -h, --help                  Display usage message\n"
         "  -V, --version               Display version string\n");
}

static void version() {
  printf("%s\n", VERSION);
}

static Cairo::RefPtr<Cairo::Surface> surface;
static int count;
static const char *output = "output";

static void newpage(CairoOutput *) {
  char *buffer;
  if(asprintf(&buffer, "%s%d.png", output, count++) < 0)
    throw std::runtime_error(std::string("asprintf: ") + strerror(errno));
  surface->write_to_png(buffer);
  free(buffer);
}

int main(int argc, char **argv) {
  try {
    int tabstop = 8;
    int width = -1, height = -1;
    double fontsize = 10.0;
    double border = 0;
    const char *font = "Courier New";
    const char *title = NULL;
    bool pageNumbering = false;
    bool italic = false;
    if(!setlocale(LC_CTYPE, ""))
      throw std::runtime_error(std::string("setlocale: ")
                               + strerror(errno));
    Pango::init();
    int opt;
    while((opt = getopt_long(argc, argv, "+hVt:w:H:f:F:o:lb:pT:i", options, NULL)) >= 0) {
      switch(opt) {
      case 't': tabstop = stringToInt(optarg); break;
      case 'w': width = stringToInt(optarg); break;
      case 'H': height = stringToInt(optarg); break;
      case 'f': fontsize = stringToDouble(optarg); break;
      case 'F': font = optarg; break;
      case 'o': output = optarg; break;
      case 'l': CairoOutput::listFonts(); return 0;
      case 'b': border = stringToDouble(optarg); break;
      case 'p': pageNumbering = true; break;
      case 'T': title = optarg; break;
      case 'i': italic = true; break;
      case 'h': help(); return 0;
      case 'V': version(); return 0;
      default: return 1;
      }
    }
    if(tabstop <= 0)
      throw std::runtime_error("invalid top stop size");
    if(width == 0 || height == 0)
      throw std::runtime_error("invalid page dimensions");
    if(fontsize <= 0)
      throw std::runtime_error("invalid font size");
    if(border < 0)
      throw std::runtime_error("invalid border size");
    Pango::FontDescription fontdesc;
    fontdesc.set_family(font);
    fontdesc.set_size(PANGO_SCALE * fontsize);
    if(width < 0 || height < 0) {
      double mw, mh;
      CairoOutput::getEmSize(Cairo::ImageSurface::create(Cairo::FORMAT_ARGB32,
                                                         1, 1),
                             fontdesc, mw, mh);
      if(width < 0) width = 80 * mw + 2 * border;
      if(height < 0) height = 66 * mh + 2 * border;
    }
    surface = Cairo::ImageSurface::create(Cairo::FORMAT_ARGB32, 
                                          width, height);
    Cairo::RefPtr<Cairo::Context> context = Cairo::Context::create(surface);
    CairoOutput o(context, fontdesc, newpage);
    o.setPageNumbering(pageNumbering);
    if(title) o.setTitle(title);
    o.setBorder(border);
    o.setItalic(italic);
    Textfile t;
    t.setTabStop(tabstop);
    if(optind >= argc) {
      t.open("-");
      t.process(o);
    } else {
      while(optind < argc) {
        t.open(argv[optind++]);
        t.process(o);
      }
    }
    o.finished();
    return 0;
  } catch(std::runtime_error &e) {
    fprintf(stderr, "ERROR: %s\n", e.what());
    return 1;
  }
}
